### PR TITLE
Add `itemSubType` argument to `InventoryType.equipments` field

### PR DESCRIPTION
**Pull request by moreal**
This pull request resolves https://github.com/planetarium/NineChronicles.Headless/issues/2189

image

### CODE REVIEW

**Review by area363 — APPROVED** [review]

**Review by longfin — APPROVED** [review]
it looks good to me, but I think that we need additional criteria about itemId, to address #2189 completely.

**Comment by moreal** [comment]
> it looks good to me, but I think that we need additional criteria about itemId, to address #2189 completely.

@longfin I couldn't imagine how they would use such APIs. I thought they can get only `Aura` subtype equipment with this API and other things can be fetched as fields (e.g., `itemId`)

Well did you imagine an API like the below one? 🤔 

```graphql
stateQuery {
	avatar(...) {
		inventory {
			equipments(itemIds: [ITEM_IDS])
		}
	}
}
```

**Comment by longfin** [comment]
> > it looks good to me, but I think that we need additional criteria about itemId, to address #2189 completely.
> 
> @longfin I couldn't imagine how they would use such APIs. I thought they can get only `Aura` subtype equipment with this API and other things can be fetched as fields (e.g., `itemId`)
> 
> Well did you imagine an API like the below one? 🤔
> 
> ```graphql
> stateQuery {
> 	avatar(...) {
> 		inventory {
> 			equipments(itemIds: [ITEM_IDS])
> 		}
> 	}
> }
> ```

Right. or, do we need a new query about inventory to add filters on `equipments`?

**Comment by moreal** [comment]
> Right. or, do we need a new query about inventory to add filters on `equipments`?

No. It will not need a new query. It can be just added. Then... please check the below cases.

1. There are items (A, B, C, D, E, F), and the `itemIds` argument was given with `[G]`
2. It should return nothing.

----

1. There are items (A[Armor], B[Sword], C[Ring], D[Armor], E[Ring], F[Ring]). The `itemIds` argument was given with `[A, F]`, and the `itemSubType` argument was given with `Aura`
2. It should return nothing. 

I realized why I feel a sense of incompatibility. I think that it means "I want to see the equipment" to give specific `itemIds`. Then other filters (`itemSubType`, `equipped`) seem unfamiliar.

What do you think about this? @longfin

**Comment by longfin** [comment]
> I realized why I feel a sense of incompatibility. I think that it means "I want to see the equipment" to give specific `itemIds`. Then other filters (`itemSubType`, `equipped`) seem unfamiliar.

Well, as with the current implementation, I guess that and-style filter might be okay. 

i.e., it's fine.

> 1. There are items (A[Armor], B[Sword], C[Ring], D[Armor], E[Ring], F[Ring]). The itemIds argument was given with [A, F], and the itemSubType argument was given with Aura
> 2. It should return nothing.

From the context of griding detection, external service will act as like as below.

```mermaid
sequenceDiagram
    actor service
    actor headless
    
    headless -->> service: Emit griding tx event
    service -->> headless: Query item info w/ equipment ids
```

When querying item info, an additional step doesn't be needed since [`Griding` action already has ids](https://github.com/planetarium/lib9c/blob/v200050/Lib9c/Action/Grinding.cs#L29)

**Comment by moreal** [comment]
@longfin I applied your idea, please review again 🙏🏻

File: NineChronicles.Headless/GraphTypes/States/Models/Item/InventoryType.cs [hunk 1]
```diff
@@ -26,15 +28,42 @@ public InventoryType()
...
+                    var itemIdsFilter = context.GetArgument<Guid[]?>("itemIds");
                     if (equippedFilter.HasValue)
                     {
                         equipments = equipments.Where(x => x.equipped == equippedFilter.Value).ToList();
                     }
+
+                    if (itemSubTypeFilter.HasValue)
+                    {
+                        equipments = equipments.Where(equipment => equipment.ItemSubType == itemSubTypeFilter.Value);
+                    }
```
Comment (longfin) [inline NineChronicles.Headless/GraphTypes/States/Models/Item/InventoryType.cs:59]:
```suggestion
                    if (itemSubTypeFilter is not null)
                    {
                        equipments = equipments.Where(equipment => equipment.ItemSubType == itemSubTypeFilter);
                    }
```

**Review by longfin — APPROVED** [review]
I left a comment related to style-consistency, but overall it seems fine.

File: NineChronicles.Headless/GraphTypes/States/Models/Item/InventoryType.cs (hunk 1 above)
Comment (moreal) [reply NineChronicles.Headless/GraphTypes/States/Models/Item/InventoryType.cs]: I applied it ✅ https://github.com/planetarium/NineChronicles.Headless/compare/80f4bb55d4c53832f65b5f2e47c0ec4e3396be02..98754d8e95417044ee290db1599ea8aa32d3961e

**Pull request merged**